internal/cli/internal/command/ssh/ssh_terminal.go

```diff
@@ -0,0 +1,204 @@
+package ssh
+
+import (
+	"context"
+	"crypto/ed25519"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"time"
+
+	sshCrypt "golang.org/x/crypto/ssh"
+
+	"github.com/briandowns/spinner"
+	"github.com/pkg/errors"
+	"github.com/superfly/flyctl/api"
+	"github.com/superfly/flyctl/helpers"
+	"github.com/superfly/flyctl/internal/client"
+	"github.com/superfly/flyctl/pkg/agent"
+	"github.com/superfly/flyctl/pkg/ssh"
+	"github.com/superfly/flyctl/terminal"
+)
+
+func spin(in, out string) context.CancelFunc {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	if !helpers.IsTerminal() {
+		fmt.Fprintln(os.Stderr, in)
+		return cancel
+	}
+
+	go func() {
+		s := spinner.New(spinner.CharSets[11], 100*time.Millisecond)
+		s.Writer = os.Stderr
+		s.Prefix = in
+		s.FinalMSG = out
+		s.Start()
+		defer s.Stop()
+
+		<-ctx.Done()
+	}()
+
+	return cancel
+}
+
+type SSHParams struct {
+	Ctx            context.Context
+	Org            *api.Organization
+	App            string
+	Dialer         agent.Dialer
+	Cmd            string
+	Stdin          io.Reader
+	Stdout         io.WriteCloser
+	Stderr         io.WriteCloser
+	DisableSpinner bool
+}
+
+func SSHConnect(p *SSHParams, addr string) error {
+	terminal.Debugf("Fetching certificate for %s\n", addr)
+
+	cert, err := singleUseSSHCertificate(p.Ctx, p.Org)
+	if err != nil {
+		return fmt.Errorf("create ssh certificate: %w (if you haven't created a key for your org yet, try `flyctl ssh establish`)", err)
```

Comment:
we also support custom error types which implement `Description() string` for more elaborate error printing. Ping me on slack and I'll guide you through that.